test/util.h

```diff
@@ -81,3 +81,23 @@ class date {
 // otherwise.
 auto get_locale(const char* name, const char* alt_name = nullptr)
     -> std::locale;
+
+// FOR_EACH macro that applies a given function to each variadic macro argument.
+// FOREACH macro helpers
+#define FOR_EACH_0(WHAT)
+#define FOR_EACH_1(WHAT, X) WHAT(X)
+#define FOR_EACH_2(WHAT, X, ...) WHAT(X), FOR_EACH_1(WHAT, __VA_ARGS__)
+#define FOR_EACH_3(WHAT, X, ...) WHAT(X), FOR_EACH_2(WHAT, __VA_ARGS__)
+#define FOR_EACH_4(WHAT, X, ...) WHAT(X), FOR_EACH_3(WHAT, __VA_ARGS__)
+#define FOR_EACH_5(WHAT, X, ...) WHAT(X), FOR_EACH_4(WHAT, __VA_ARGS__)
+#define FOR_EACH_6(WHAT, X, ...) WHAT(X), FOR_EACH_5(WHAT, __VA_ARGS__)
+#define FOR_EACH_7(WHAT, X, ...) WHAT(X), FOR_EACH_6(WHAT, __VA_ARGS__)
+#define FOR_EACH_8(WHAT, X, ...) WHAT(X), FOR_EACH_7(WHAT, __VA_ARGS__)
+#define FOR_EACH_9(WHAT, X, ...) WHAT(X), FOR_EACH_8(WHAT, __VA_ARGS__)
+
+#define FOR_EACH_GET_MACRO(_1, _2, _3, _4, _5, _6, _7, _8, _9, NAME, ...) NAME
+#define FOR_EACH(action, ...)                                           \
+  FOR_EACH_GET_MACRO(__VA_ARGS__ __VA_OPT__(, ) FOR_EACH_9, FOR_EACH_8, \
+                     FOR_EACH_7, FOR_EACH_6, FOR_EACH_5, FOR_EACH_4,    \
+                     FOR_EACH_3, FOR_EACH_2, FOR_EACH_1, FOR_EACH_0)    \
+  (action __VA_OPT__(, ) __VA_ARGS__)
```

Comment:
Let's not introduce this and avoid new macros in tests.

Reply:
In this case getting rid of the macros that means duplicating all the tests when testing for a format function that returns an `std::string_view`. I'm not sure which of two evils is better.

AFAIK, a constexpr function returning the result of a format as `std::string_view` necessarily has to take all arguments as NTTP, since it needs to store the formatted result as a `static constexpr`, i.e. one such constant per unique set of arguments. And the issue there is that we can't pass `"bla"` as NTTP, it needs to be some kind of `fixed_string`.

I can't figure out a way to do this conditionally per variadic NTTP. As soon as the compiler sees a `const char[]` it's game over. Hence why I did the conversion of `const char[]` by wrapping each template argument in a lambda call with this `FOR_EACH` macro.

Would you prefer I duplicate some/all test cases?

Reply:
Nevermind the above, I figured out a way to get rid of the `FOR_EACH`.

Reply:
Also I don't think we need those `string_view` / NTTP tests since they are testing the same logic.

Reply:
I would agree, except I'm seeing different compile errors with the NTTP tests vs the existing ones (mainly missing `constexpr` specifiers).

Reply:
Could you provide an example on godbolt?

Reply:
Sure: https://godbolt.org/z/eqorr8qWr. Note that there's a bunch of things this example doesn't support. It's just a MWE of one of the differences I saw.

Of course this could be due to something I implemented wrongly. E.g. formatting `42` uses a different call chain than `uint64_t{42}`, and I might have been going down a bunch of paths trying to make everything work with NTTP. Might have confused myself at some point :smile:.

Reply:
Thanks for the example. Looks like the only thing that is missing from the coverage is `formatted_size` (https://godbolt.org/z/8411Eaoh3) so let's revert all the test changes here and add a call to it to `test_format`, something like:

```
  constexpr auto n = fmt::formatted_size(format, args...);
```

Reply:
Done. Also added a missing test case to catch the missing `constexpr` in the `FMT_FORMAT_AS` macro.